packages/react/src/ReactBlockSpec.tsx

```diff
@@ -76,57 +120,57 @@ export function createReactBlockSpec<
       return parse(blockConfig);
     },
 
-    renderHTML({ HTMLAttributes }) {
-      return render(blockConfig, HTMLAttributes);
-    },
-
     addNodeView() {
-      const BlockContent: FC<NodeViewProps> = (props: NodeViewProps) => {
-        const Content = blockConfig.render;
-
-        // Add props as HTML attributes in kebab-case with "data-" prefix
-        const htmlAttributes: Record<string, string> = {};
-        for (const [attribute, value] of Object.entries(props.node.attrs)) {
-          if (attribute in blockConfig.propSchema) {
-            htmlAttributes[camelToDataKebab(attribute)] = value;
-          }
-        }
+      return (props) =>
+        ReactNodeViewRenderer(
+          (props: NodeViewProps) => {
+            // Gets the BlockNote editor instance
+            const editor = this.options.editor as BlockNoteEditor<BSchema>;
+            // Gets the block
+            const block = getBlockFromPos<BType, PSchema, BSchema>(
+              props.getPos,
+              editor,
+              this.editor,
+              blockConfig.type
+            );
 
-        // Gets BlockNote editor instance
-        const editor = this.options.editor!;
-        // Gets position of the node
-        const pos =
-          typeof props.getPos === "function" ? props.getPos() : undefined;
-        // Gets TipTap editor instance
-        const tipTapEditor = editor._tiptapEditor;
-        // Gets parent blockContainer node
-        const blockContainer = tipTapEditor.state.doc.resolve(pos!).node();
-        // Gets block identifier
-        const blockIdentifier = blockContainer.attrs.id;
-        // Get the block
-        const block = editor.getBlock(blockIdentifier)!;
-        if (block.type !== blockConfig.type) {
-          throw new Error("Block type does not match");
-        }
-
-        return (
-          <NodeViewWrapper
-            className={blockStyles.blockContent}
-            data-content-type={blockConfig.type}
-            {...htmlAttributes}>
-            <Content block={block} editor={editor} />
-          </NodeViewWrapper>
-        );
-      };
+            const BlockContent = reactRenderWithBlockStructure<
+              BType,
+              PSchema,
+              ContainsInlineContent,
+              BSchema
+            >(blockConfig.render, block, this.options.editor);
 
-      return ReactNodeViewRenderer(BlockContent, {
-        className: blockStyles.reactNodeViewRenderer,
-      });
+            return <BlockContent />;
+          },
+          {
+            className: blockStyles.reactNodeViewRenderer,
+          }
+        )(props);
     },
   });
 
   return {
     node: node,
     propSchema: blockConfig.propSchema,
+    serialize: (block, editor) => {
+      // TODO: This doesn't seem like a great way of doing things - any better
```

Comment:
why do you think this? isn't this exactly the purpose of `renderToString`?

Reply:
Haven't really used `renderToString` before tbh, in general converting React to HTML feels a bit like something you shouldn't do. But if this is actually what it's supposed to be for then ig it's not a problem here.